### DISCUSSION TITLE
Connection establishment b/w Multimaster cluster

### DISCUSSION THREAD

**Discussion by MohamedShaj** [k3s-io/k3s#6456] [Q&A]
**Environmental Info:**
K3s Version: 
latest one

Node(s) CPU architecture, OS, and Version: 
any os and specification

Cluster Configuration:
2 master cluster

**Describe the bug:**
My query is, I have two master node clusters (where master and agent installed by default in each node)

- So I need to understand technically, how the two masters form as a cluster
- what services are interconnected b/w the two master nodes 
- How securely it is connected and what ports are used to establish the connection?
- how to figure out which is the active master (b/w the two masters) because if we have multi-master also only one master going to be active and other going to be standby

I have just made a architecture diagram by reference to https://github.com/rancher/docs/issues/3268 , may I know what is supervisor in it ? 

And please answer my queries with respect to below dia ?

![ticket](https://user-images.githubusercontent.com/102578451/200278653-8bf3dca4-66ec-4c0f-88e9-179580e5cacf.jpg)

**Comment by brandond** [comment]
A two-server cluster is not supported when using etcd. You must have an odd number of nodes in order to meet the etcd quorum requirements.
https://etcd.io/docs/v3.5/faq/#why-an-odd-number-of-cluster-members

**Comment by MohamedShaj** [comment]
okay now it is in odd number, how the connection is established from master to other masters ?
How securely it is connected and what ports are used to establish the connection?
how to figure out which is the active master (b/w the 3 masters) because if we have multi-master also only one master going to be active and other going to be standby
I have just made a architecture diagram by reference to https://github.com/rancher/docs/issues/3268 , may I know what is supervisor in it ?

And please answer my queries with respect to below dia ?


![ticket (3)](https://user-images.githubusercontent.com/102578451/200375135-8728cdf1-6750-494f-b213-d296eca832d4.jpg)

**Comment by brandond** [comment]
You can find the required ports for K3s at https://docs.k3s.io/installation/requirements#networking

Also note that K3s does not have masters. We have servers and agents.

Etcd and each of the Kubernetes control-plan components have a single leader, but you shouldn't really concern yourself with node is currently the leader. Etcd manages leader election internally via RAFT, and the individual Kubernetes controllers use take leases through the Kubernetes apiserver to determine which one is active. The Kubernetes control-plane components do not communicate directly to each other; everything is done via the apiserver's etcd datastore. For further documentation on those components, consult the relevant upstream project docs.

**Comment by MohamedShaj** [comment]
> You can find the required ports for K3s at https://docs.k3s.io/installation/requirements#networking
> 
> Also note that K3s does not have masters. We have servers and agents.
> 
> Etcd and each of the Kubernetes control-plan components have a single leader, but you shouldn't really concern yourself with node is currently the leader. Etcd manages leader election internally via RAFT, and the individual Kubernetes controllers use take leases through the Kubernetes apiserver to determine which one is active. The Kubernetes control-plane components do not communicate directly to each other; everything is done via the apiserver's etcd datastore. For further documentation on those components, consult the relevant upstream project docs.

- Thanks for the clarity servers and masters are different, so ETCD manages the leader using RAFT method and that node is considered as a leader and that control plane alone going to manage the cluster and other two going to standby ?
- This is all about ETCD, if we are going with mySQL/PostgreSQL/SQLite how the the things will be changed ?
- can you share me relevant upstream project docs link or any whitepaper on how the connectivity happening in back ground with multi master ??

Because i knew tunnel proxy using in front of kube proxy with bidirection usage (with reverse proxy method) asper  https://github.com/rancher/remotedialer, so how the tunnel proxy will be handled for the multi master cluster ??

On these stuffs, there is no in-depth architecture diagram so curious to know on this !!!!

**Reply by brandond** [reply]
> Thanks for the clarity servers and masters are different, so ETCD manages the leader using RAFT method and that node is considered as a leader and that control plane alone going to manage the cluster and other two going to standby ?

No, they are completely separate. The etcd cluster has a leader, but it is opaque to clients which node that is, although you can use etcdctl to find out if you care for some reason. Each Kubernetes component has a single replica that is the leader; you can use `kubectl get lease` to figure out which one is the leader for that component. Each of these is completely independent from all the others.

> This is all about ETCD, if we are going with mySQL/PostgreSQL/SQLite how the the things will be changed ?

Same thing, but without etcd.

> can you share me relevant upstream project docs link or any whitepaper on how the connectivity happening in back ground with multi master ??

Etcd docs are at https://etcd.io/docs/v3.5/. 

> Because i knew tunnel proxy using in front of kube proxy with bidirection usage (with reverse proxy method)

This only concerns communication between the kubelet and apiserver. Etcd servers talk to each other to replicate data, apiservers talk to etcd, everything else talks to an apiserver.

**Reply by MohamedShaj** [reply]
Hi Davidson,

Thank you for all the quick answers,only one question still left, 


- in 3 server node cluster , if any one server is going down ...how the HA will be working here ? 
- in k8s if we have 3master node cluster ...only one master ll be active and other ll be in standby...so similar behaviour is there in k3s and how the master node failure is taken care by HA in k3s ?

Thanks in advance 

Have a nice day

**Reply by brandond** [reply]
> in 3 server node cluster , if any one server is going down ...how the HA will be working here ?

HA of what component?

> in k8s if we have 3master node cluster ...only one master ll be active and other ll be in standby...so similar behaviour is there in k3s and how the master node failure is taken care by HA in k3s ?

No, as I mentioned above, each component handles leader election individually. Each node may be the leader for a different component at a given time. There is no coordination between components to ensure that one node is leader of everything and the others are standby. If you care for some reason (and generally, you shouldn't) you would have to look at each component to figure out what node is active.

**Reply by MohamedShaj** [reply]
> > in 3 server node cluster , if any one server is going down ...how the HA will be working here ?
> 
> HA of what component?
> 
> > in k8s if we have 3master node cluster ...only one master ll be active and other ll be in standby...so similar behaviour is there in k3s and how the master node failure is taken care by HA in k3s ?
> 
> No, as I mentioned above, each component handles leader election individually. Each node may be the leader for a different component at a given time. There is no coordination between components to ensure that one node is leader of everything and the others are standby. If you care for some reason (and generally, you shouldn't) you would have to look at each component to figure out what node is active.

I was talking about HA of Control plane (server) and how the leader is selected and how it is handled when it is going down in k3s ?

**Reply by brandond** [reply]
> I was talking about HA of Control plane (server) and how the leader is selected and how it is handled when it is going down in k3s ?

There are many components to the control-plane - the apiserver, kube-scheduler, kube-controller-manager, cloud-controller-manager, and so on. The apiservers are just a web service, and do not perform leader election. The remaining components all use leases. This is not really something you need to worry about, but there are blog posts available on the topic if you want to read up on it. It is not something unique to K3s. 
* https://github.com/kubernetes/client-go/blob/master/examples/leader-election/main.go
* https://medium.com/hybrid-cloud-hobbyist/leader-election-architecture-kubernetes-32600da81e3c

**Reply by MohamedShaj** [reply]
still i have a confusion, if i have 3 machines and server & agent as a single process is running in all 3 machine, now I have 3server and 3 agents,
 these are my assumption, please let me know this is right assumption,

- All 3 API server will active because it is a web services and do not come inside the lease election
-  All 3 ETCD forms cluster and have a leader for it and the other 2 will act as the replica (asper RAFT mechanism)
-  scheduler, controller and so on will be under lease mechanism, so only the leader will be in active state and another node scheduler, and controller will be in standby 


one more thing https://github.com/rancher/docs/issues/3268 in this what service is pointed as supervisor ??
any other things to be added please let me know,

(actually that we should not worry about the leader election, but for our application failover methods must to be cleared because we have some application dependency on this architecture)

**Reply by brandond** [reply]
The supervisor is just the provisioning and configuration logic custom to K3s. It handles generating certificates and configuration files, managing etcd cluster membership, and so on. It is not technically part of the control-plane, as it manages the configuration control-plane itself. It is mostly only used when k3s is starting up.

**Reply by MohamedShaj** [reply]
![image](https://user-images.githubusercontent.com/102578451/200811420-bed9ef47-adb7-486c-b5f5-3c7889fd9696.png)

we can take the above picture as an example, why tunnel proxy is sending request to supervisor, logically we need tunnel proxy from k3s server right to receive the request from  agent tunnel proxy,
